lib/state.ml

```diff
@@ -23,85 +23,150 @@ let find_or_add_repo' state repo_url =
+    let update_pipeline_status =
+      update_builds_in_branches ~branches:n.branches
+        ~default_builds_map:(StringMap.singleton build_number init_build_state) ~f:(fun builds_map ->
+          let updated_status = update_build_status builds_map build_number in
+          StringMap.add build_number updated_status builds_map)
+    in
+    let rm_successful_build =
+      update_builds_in_branches ~branches:n.branches ~f:(fun builds_map ->
+          StringMap.remove build_number builds_map
+          |> StringMap.filter (fun build_number' build_status ->
+                 (* Remove old builds without failed steps because they were fixed and cleaned from state *)
+                 match build_status.State_t.failed_steps with
+                 | [] when build_number' < build_number && build_status.is_finished -> false
+                 | _ -> true))
+    in
+    let rm_successful_step =
+      update_builds_in_branches ~branches:n.branches ~f:(fun builds_map ->
+          StringMap.mapi
+            (fun build_number' (build_status : State_t.build_status) ->
+              let failed_steps =
+                List.filter
+                  (* remove the fixed step from previous finished builds *)
+                    (fun (s : State_t.failed_step) ->
+                    not (build_number' < build_number && s.name = n.context && build_status.is_finished))
+                  build_status.failed_steps
+              in
+              { build_status with failed_steps })
+            builds_map)
+    in
+    let repo_state = find_or_add_repo' state n.repository.url in
+    (match n.state with
+    | Success ->
+      (* If a build step is successful, we remove it from the failed steps list of past builds.
+         If old builds have no more failed steps, we remove them.
+         If the whole build is successful, we remove it from state to avoid the state file on disk growing too much.
+      *)
+      (match is_pipeline_step with
+      | true ->
+        repo_state.pipeline_statuses <- StringMap.update pipeline_name rm_successful_step repo_state.pipeline_statuses
+      | false ->
+        repo_state.pipeline_statuses <- StringMap.update pipeline_name rm_successful_build repo_state.pipeline_statuses)
+    | Error | Failure | Pending ->
+      repo_state.pipeline_statuses <- StringMap.update pipeline_name update_pipeline_status repo_state.pipeline_statuses)
 
-let set_repo_pipeline_commit { state } repo_url ~pipeline ~commit =
+let set_repo_pipeline_commit { state } (n : Github_t.status_notification) =
   let rotation_threshold = 1000 in
-  let repo_state = find_or_add_repo' state repo_url in
+  let repo_state = find_or_add_repo' state n.repository.url in
+  let pipeline =
+    (* We only need to track messages for the base pipeline, not the steps *)
+    match Util.Build.parse_context ~context:n.context ~build_url:(Option.default "" n.target_url) with
```

Comment:
iiuc calling the function with `build_url` as empty string will never succeed

Reply:
it shouldn't happen in buildkite, but for other CIs this value might be empty. See [this case](https://github.com/ahrefs/monorobot/blob/master/mock_payloads/status.success_public_repo_no_buildkite.json#L5).

So here we protect against it.

Reply:
> creates more assumptions for the callee to manage

Makes sense. Good call, thanks